### PR TITLE
Implement `where`

### PR DESCRIPTION
**Pull request by Pranavchiku**
#1312

```fortran
subroutine where_1
    implicit none
    integer a(10)
    integer b(10)
    
    where (a > b) 
        a = 1
    endwhere
end subroutine

program main
call where_1
end program
```

<details>
<summary>ASR</summary>

<p>

```clojure
(TranslationUnit
    (SymbolTable
        1
        {
            main:
                (Program
                    (SymbolTable
                        3
                        {
                            
                        })
                    main
                    []
                    [(SubroutineCall
                        1 where_1
                        ()
                        []
                        ()
                    )]
                ),
            where_1:
                (Function
                    (SymbolTable
                        2
                        {
                            a:
                                (Variable
                                    2
                                    a
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Integer 4 [((IntegerConstant 1 (Integer 4 []))
                                    (IntegerConstant 10 (Integer 4 [])))])
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            b:
                                (Variable
                                    2
                                    b
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Integer 4 [((IntegerConstant 1 (Integer 4 []))
                                    (IntegerConstant 10 (Integer 4 [])))])
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    where_1
                    (FunctionType
                        []
                        ()
                        Source
                        Implementation
                        ()
                        .false.
                        .false.
                        .false.
                        .false.
                        .false.
                        []
                        []
                        .false.
                    )
                    []
                    []
                    [(Where
                        (IntegerCompare
                            (Var 2 a)
                            Gt
                            (Var 2 b)
                            (Logical 4 [((IntegerConstant 1 (Integer 4 []))
                            (IntegerConstant 10 (Integer 4 [])))])
                            ()
                        )
                        [(=
                            (Var 2 a)
                            (IntegerConstant 1 (Integer 4 []))
                            ()
                        )]
                        []
                    )]
                    ()
                    Public
                    .false.
                    .false.
                )
        })
    []
)
```

</>

</details>

### CODE REVIEW

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 1]
```diff
@@ -2318,6 +2318,113 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
+                                left_array_item, cmp->m_op, right, cmp->m_type, nullptr));
+        }
+        Vec<ASR::stmt_t*> if_body;
+        if_body.reserve(al, x.n_body);
+        tmp = ASR::make_If_t(al, loc, test_new, if_body.p,
+                            if_body.size(), nullptr, 0);
+        doloop_body.push_back(al, ASRUtils::STMT(tmp));
+
+        tmp = ASR::make_DoLoop_t(al, loc, nullptr, head, doloop_body.p, doloop_body.size());
+    }
```
Comment (certik) [inline src/lfortran/semantics/ast_body_visitor.cpp:2426]: In this function we should simply construct an ASR where stmt node, representing this feature.

Then we can write an ASR->ASR pass that expresses it using loops.

Comment (Pranavchiku) [reply src/lfortran/semantics/ast_body_visitor.cpp:2331]: Yes, perfect.

**Comment by Pranavchiku** [comment]
Failing test: `CI / Test integration_tests with GFortran (pull_request)` is not related to this PR. I tested on locally it works with GFortran, I guess it will get resolve on rerun.

**Comment by Pranavchiku** [comment]
I have a naive design for ASR pass ready. Will open PR once this gets merged.

**Review by certik — APPROVED** [review]
Looks great!

**Pull request merged**